File: Wave Technical Challege/Wave Technical Challege/ViewController.m
//
//  ViewController.m
//  Wave Technical Challege
//
//  Created by Binoli Mhatre on 2019-03-18.
//  Copyright © 2019 Binoli Mhatre. All rights reserved.
//

#import "ViewController.h"
#import "ProductCell.h"
#define kWaveURL @"https://api.waveapps.com/businesses/89746d57-c25f-4cec-9c63-34d7780b044b/products/"
#define kGoogleURL @"https://www.google.com/"
#define kCellID @"ProductCellID"
#define kProductName @"name"
#define  kProductPrice @"price"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    _activityIndicator.center =self.view.center;
    [_activityIndicator startAnimating ];
    [_lblStatus setHidden:true];
    [_productsTable setHidden:true];
    CGFloat statusBarY =[[UIApplication sharedApplication]statusBarFrame].size.height;
    CGRect frame=CGRectMake(0, statusBarY+1, self.view.frame.size.width, self.view.frame.size.height-(statusBarY+1));
    _productsTable.frame=frame;
    _lblStatus.frame=frame;
    
    //Proceed if Internet is avail
    if([self isInternetAvail])
    {
        //Internet available. Load products
        [_productsTable setHidden:false];
        [self callWebService];
    }
    else
    {
        //No Internet. Display status and do nothing
        
        //Stop ActivityIndicator and Update Status
        [_activityIndicator stopAnimating];
        [_activityIndicator setHidden:true];
        [_productsTable setHidden:true];
        [_lblStatus setHidden:false];
        _lblStatus.text=@"Internet Not Available!";
        _lblStatus.textAlignment =NSTextAlignmentCenter;
    }
}

#pragma mark - UITableViewDatasource Methods
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [productsArray count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *cellID = kCellID;
    ProductCell *cell = (ProductCell *)[tableView dequeueReusableCellWithIdentifier:cellID];
    //Get Product Name
    cell.lblProductName.text = [[productsArray objectAtIndex:indexPath.row] valueForKey:kProductName];
    //lblPrice set X coordinate to match different screen sizes
    CGFloat x= self.view.frame.size.width - cell.lblPrice.frame.size.width;
    CGRect priceFrame=CGRectMake(x, cell.lblPrice.frame.origin.y, cell.lblPrice.frame.size.width, cell.lblPrice.frame.size.height);
    cell.lblPrice.frame=priceFrame;
    //Get Product Price
    NSNumber *price =[[productsArray objectAtIndex:indexPath.row] valueForKey:kProductPrice];
    cell.lblPrice.text = [NSString stringWithFormat:@"$ %@",[price stringValue]];
    
    return cell;
}

#pragma mark - UITableViewDelegate Method
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 100.0f; //CustomCell height
}


#pragma mark - Internet and WebService Methods
-(BOOL)isInternetAvail
{
    NSURL *googleURL = [NSURL URLWithString:kGoogleURL];
    NSData *data = [NSData dataWithContentsOfURL:googleURL];
    if (data)
    {
        return TRUE; //Available
    }
    return FALSE; //Notavailable
}


-(void) callWebService
{
    NSMutableURLRequest *request = [NSMutableURLRequest new];
     //set request parameters
    request.HTTPMethod = @"GET";
    [request setURL:[NSURL URLWithString:kWaveURL]];
    [request setValue:@"Bearer 6W9hcvwRvyyZgPu9Odq7ko8DSY8Nfm" forHTTPHeaderField:@"Authorization"];
    // NSURLSession
    NSURLSessionConfiguration *sessionConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:sessionConfig delegate:nil  delegateQueue:[NSOperationQueue mainQueue]];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data,  NSURLResponse * _Nullable response, NSError * _Nullable error)
    {
    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
    
    //Stop ActivityIndicator
    [_activityIndicator stopAnimating];
    [_activityIndicator setHidden:true];
        
    if(httpResponse.statusCode==200)
    {//Success
        NSArray *responseArray = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
        //NSLog(@"Response array: %@", responseArray);
        //Display products
        productsArray =[[ NSMutableArray alloc]initWithArray:responseArray];
        [_productsTable reloadData];
    }
    else
    {//Fail
        //NSLog(@"Request Failed with status code %ld",(long)httpResponse.statusCode);
        
        //Update Status
        [_productsTable setHidden:true];
         [_lblStatus setHidden:false];
        _lblStatus.text=@"Could not retreive products!";
        _lblStatus.textAlignment =NSTextAlignmentCenter;
    }
    
    }];
    [task resume];
}


#pragma mark - MemoryWarning method

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: WaveChallenge/WaveChallenge/ViewController.swift
//
//  ViewController.swift
//  WaveChallenge
//
//  Created by Chris Ta on 2018-02-25.
//  Copyright © 2018 ChrisHuyTa. All rights reserved.
//

import UIKit
import Alamofire

private let business_id = "89746d57-c25f-4cec-9c63-34d7780b044b"
private let access_token = "6W9hcvwRvyyZgPu9Odq7ko8DSY8Nfm"
private let productAPIUrl = "https://api.waveapps.com/businesses/"

class ViewController: UIViewController {

    let productListView: UITableView = {
        let tableview = UITableView()
        tableview.translatesAutoresizingMaskIntoConstraints = false
        return tableview
    }()
    
    var products = [[String: Any]]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        productListView.dataSource = self
        
        self.view.addSubview(productListView)
        setupProductListView()
        
        fetchProducts()
    }
    
    //fet
    func fetchProducts() {
        let url: URL = URL(string: productAPIUrl + business_id + "/products/")!
        let headers: HTTPHeaders = ["Authorization": "Bearer " + access_token]
        Alamofire.request(url, method: .get, parameters: nil, encoding: URLEncoding.queryString, headers: headers)
            .validate()
            .responseJSON { (response) in
                if let json = response.result.value {
                    self.products = json as! [[String:Any]]
                    self.productListView.reloadData()
                }
                
        }
    }
    
    func setupProductListView() {
        productListView.leftAnchor.constraint(equalTo: self.view.leftAnchor).isActive = true
        productListView.topAnchor.constraint(equalTo: self.view.topAnchor).isActive = true
        productListView.rightAnchor.constraint(equalTo: self.view.rightAnchor).isActive = true
        productListView.bottomAnchor.constraint(equalTo: self.view.bottomAnchor).isActive = true
    }

}

extension ViewController: UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return products.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell.init(style: .value1, reuseIdentifier: "wave")
        var product = products[indexPath.row]
        cell.textLabel?.text = product["name"] as? String
        let price = formatCurrency(amount: product["price"] as! Float)
        cell.detailTextLabel?.text = "\(price)"
        cell.detailTextLabel?.textColor = UIColor.blue
        
        return cell
    }
    
    func formatCurrency(amount: Float) -> String {
        let currencyFormatter = NumberFormatter()
        currencyFormatter.usesGroupingSeparator = true
        currencyFormatter.numberStyle = .currency
        currencyFormatter.currencyCode = "CAD"
        return currencyFormatter.string(from: NSNumber(value: amount))!
    }
}

